lib/index.js

```diff
@@ -393,7 +401,7 @@ Mongoose.prototype.model = function(name, schema, collection, skipInit) {
   }
 
   var connection = options.connection || this.connection;
-  model = this.Model.compile(name, schema, collection, connection, this);
+  model = this.Model.compile(model || name, schema, collection, connection, this);
```

Comment:
Not sure where the problem you see is, can you elaborate? `model` is unset if the `name` passed in is a string, which means that `Model.compile` receives a string name rather than an ES6 class, and everything proceeds normally.